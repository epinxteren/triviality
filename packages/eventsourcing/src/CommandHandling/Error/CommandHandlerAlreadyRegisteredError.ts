import { CommandConstructor } from '../Command';
import { ClassUtil } from '../../ClassUtil';
import { CommandHandler } from '../CommandHandler';

export class CommandHandlerAlreadyRegisteredError extends Error {

  public static alreadyRegistered(handler: CommandHandler, handlerFunction: string, command: CommandConstructor) {
    const handlerName = ClassUtil.nameOffInstance(handler);
    const commandName = ClassUtil.nameOffInstance(command);
    return new this(
      `${commandName} cannot be registered on ${handlerName}.${handlerFunction} because already registered to other handler`);
  }

}
