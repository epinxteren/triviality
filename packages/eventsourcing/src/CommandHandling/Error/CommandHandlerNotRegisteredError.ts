import { Command } from '../Command';
import { ClassUtil } from '../../ClassUtil';

export class CommandHandlerNotRegisteredError extends Error {

  public static missingHandler(command: Command) {
    const commandName = ClassUtil.nameOffInstance(command);
    return new this(`Missing command handler registered for ${commandName}`);
  }

}
