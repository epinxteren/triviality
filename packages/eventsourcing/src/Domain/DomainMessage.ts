import { DomainEvent } from './DomainEvent';
import { ClassUtil } from '../ClassUtil';
import { Identity } from '../ValueObject/Identity';

export class DomainMessage<DM extends DomainEvent = DomainEvent, Id extends Identity = Identity> {

  public static recordNow<DM extends DomainEvent = DomainEvent, Id extends Identity = Identity>(
    aggregateId: Id,
    playhead: number,
    payload: DM,
    metadata: { [key: string]: any } = {}): DomainMessage {
    return new DomainMessage(aggregateId, playhead, payload, new Date(), metadata);
  }

  constructor(public readonly aggregateId: Id,
              public readonly playhead: number,
              public readonly payload: DM,
              public readonly recordedOn: Date,
              public readonly metadata: { [key: string]: any } = {}) {
  }

  public toString(): string {
    const name = ClassUtil.nameOffInstance(this.payload);
    return `${this.aggregateId}:${this.playhead}:${name}`;
  }

}
