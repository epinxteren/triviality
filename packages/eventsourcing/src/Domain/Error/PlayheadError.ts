export class PlayheadError extends Error {

  public static create(expected: number, actual: number) {
    return new PlayheadError(`Playhead does not match expected ${expected} given ${actual}`);
  }

}
