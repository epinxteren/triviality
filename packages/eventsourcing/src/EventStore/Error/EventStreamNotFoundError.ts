import { Identity } from '../../ValueObject/Identity';

export class EventStreamNotFoundError extends Error {
  public static streamNotFound(id: Identity) {
    return new this(`EventStream not found for aggregate with id ${id.toString()}`);
  }
}
