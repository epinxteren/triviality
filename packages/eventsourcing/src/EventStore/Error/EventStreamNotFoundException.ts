import { Identity } from '../../ValueObject/Identity';

export class EventStreamNotFoundException extends Error {
  public static streamNotFound(id: Identity) {
    return new this(`EventStream not found for aggregate with id ${id.toString()}`);
  }
}
