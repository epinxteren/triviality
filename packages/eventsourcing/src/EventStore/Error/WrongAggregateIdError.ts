import { Identity } from '../../ValueObject/Identity';

export class WrongAggregateIdError extends Error {

  public static create(actualId: Identity, expectedId: Identity) {
    return new WrongAggregateIdError(`Wrong aggregate id ${actualId} is not expected ${expectedId}`);
  }

}
