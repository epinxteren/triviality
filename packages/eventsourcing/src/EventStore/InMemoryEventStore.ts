import { concatMap, filter, toArray } from 'rxjs/operators';
import { MutexInterface } from 'async-mutex';
import { from } from 'rxjs';
import { ValueMutex } from '../ValueMutex';
import { WrongAggregateIdError } from './Error/WrongAggregateIdError';
import { Mutex } from 'async-mutex';
import { EventStore } from './EventStore';
import { EventStreamNotFoundError } from './Error/EventStreamNotFoundError';
import { DomainEventStream } from '../Domain/DomainEventStream';
import { Identity } from '../ValueObject/Identity';
import { DomainMessage } from '../Domain/DomainMessage';
import { PlayheadError } from '../Domain/Error/PlayheadError';

export class InMemoryEventStore<Id extends Identity = Identity>
  implements EventStore<Id> {
  public static fromArray<Id extends Identity = Identity>(events: DomainMessage[], mutexFactory: () => MutexInterface = () => new Mutex()) {
    const instance = new this<Id>(mutexFactory);
    instance.events = new ValueMutex([...events], mutexFactory());
    const mapped = new Map<Id, DomainMessage[]>();
    for (const event of events) {
      const list = mapped.get(event.aggregateId as Id) || [];
      list.push(event);
      mapped.set(event.aggregateId as Id, list);
    }
    mapped.forEach((value, key) => {
      instance.eventsMap.set(key, new ValueMutex(value, mutexFactory()));
    });
    return instance;
  }

  protected readonly eventsMap: Map<Id, ValueMutex<DomainMessage[]>> = new Map();
  protected events = new ValueMutex<DomainMessage[]>([], this.mutexFactory());

  constructor(private mutexFactory: () => MutexInterface = () => new Mutex()) {
  }

  public async has(id: Id): Promise<boolean> {
    const map = this.eventsMap.get(id);
    if (!map) {
      return false;
    }
    return map.runExclusive((value) => {
      return value.length !== 0;
    });
  }

  public load(id: Id): DomainEventStream {
    const map = this.eventsMap.get(id);
    if (!map) {
      throw EventStreamNotFoundError.streamNotFound(id);
    }
    return from(map.runExclusive((events) => {
      if (events.length !== 0) {
        return events;
      }
      throw EventStreamNotFoundError.streamNotFound(id);
    })).pipe(concatMap((events) => events));
  }

  public loadAll(): DomainEventStream {
    return from(this.events.runExclusive((events) => events)).pipe(concatMap((events) => events));
  }

  public loadFromPlayhead(id: Id, playhead: number): DomainEventStream {
    return this.load(id).pipe(filter((message: DomainMessage) => playhead <= message.playhead));
  }

  /**
   * This save the hole stream or nothing at all.
   */
  public async append(id: Id, eventStream: DomainEventStream): Promise<void> {
    if (!this.eventsMap.has(id)) {
      this.eventsMap.set(id, new ValueMutex<DomainMessage[]>([], this.mutexFactory()));
    }
    await this.eventsMap.get(id)!.runExclusive(async (target) => {
      await this.events.runExclusive(async (all) => {
        let lastPlayhead = this.lastPlayheadOf(target);
        const events = await eventStream.pipe(
          concatMap(async (event: DomainMessage) => {
            lastPlayhead += 1;
            if (event.aggregateId !== id) {
              throw WrongAggregateIdError.create(event.aggregateId, id);
            }
            if (event.playhead !== lastPlayhead) {
              throw PlayheadError.create(lastPlayhead, event.playhead);
            }
            return event;
          }),
          toArray(),
        ).toPromise();
        // Add if everything is ok.
        target.push(...events);
        all.push(...events);
        return events;
      });
    });
  }

  private lastPlayheadOf(events: DomainMessage[]): number {
    return events.length !== 0 ? events[events.length - 1].playhead : -1;
  }
}
