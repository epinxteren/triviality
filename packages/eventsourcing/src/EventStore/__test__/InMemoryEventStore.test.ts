import 'jest';
import { PlayheadError } from '../../Domain/Error/PlayheadError';
import { ScalarIdentity } from '../../ValueObject/ScalarIdentity';
import { InMemoryEventStore } from '../InMemoryEventStore';
import { EventStreamNotFoundError } from '../Error/EventStreamNotFoundError';
import { SimpleDomainEventStream } from '../../Domain/SimpleDomainEventStream';
import { DomainMessage } from '../../Domain/DomainMessage';
import { map, take, toArray } from 'rxjs/operators';
import { Mutex } from 'async-mutex';
import { timer } from 'rxjs';
import v4 from 'uuid/v4';

class DomainEvent {

}

describe('InMemoryEventStore', () => {
  const date = new Date();
  date.setTime(1535183762);

  it('Throws an exception loading an unknown event stream', async () => {
    const repository = new InMemoryEventStore();
    const id = new ScalarIdentity('38459347598437');
    const eventStreamNotFoundException = new EventStreamNotFoundError(
      'EventStream not found for aggregate with id 38459347598437');
    return expect(() => {
      repository.load(id);
    }).toThrowError(eventStreamNotFoundException as any);
  });

  it('Can append domain event', async () => {
    const repository = new InMemoryEventStore();
    const id = new ScalarIdentity('38459347598437');
    const event = new DomainMessage(id, 0, new DomainEvent(), date);
    const eventstream = SimpleDomainEventStream.of([event]);
    await repository.append(id, eventstream);
    expect(await repository.load(id).pipe(toArray()).toPromise()).toEqual([event]);
  });

  it('Can append multiple domain events at once', async () => {
    const repository = new InMemoryEventStore();
    const id = new ScalarIdentity('38459347598437');
    const event1 = new DomainMessage(id, 0, new DomainEvent(), date);
    const event2 = new DomainMessage(id, 1, new DomainEvent(), date);
    const eventStream = SimpleDomainEventStream.of([event1, event2]);
    await repository.append(id, eventStream);
    expect(await repository.load(id).pipe(toArray()).toPromise()).toEqual([event1, event2]);
  });

  it('Can append multiple domain events streams', async () => {
    const repository = new InMemoryEventStore();
    const id = new ScalarIdentity('38459347598437');
    const event1 = new DomainMessage(id, 0, new DomainEvent(), date);
    const event2 = new DomainMessage(id, 1, new DomainEvent(), date);
    const eventstream1 = SimpleDomainEventStream.of([event1, event2]);
    await repository.append(id, eventstream1);
    const event3 = new DomainMessage(id, 2, new DomainEvent(), date);
    const event4 = new DomainMessage(id, 3, new DomainEvent(), date);
    const eventstream2 = SimpleDomainEventStream.of([event3, event4]);
    await repository.append(id, eventstream2);
    expect(await repository.load(id).pipe(toArray()).toPromise()).toEqual([event1, event2, event3, event4]);
  });

  it('Can have multiple different events streams', async () => {
    const repository = new InMemoryEventStore();
    const aggregate1Id = new ScalarIdentity('aggregate1 ID');
    const event1Aggregate1 = new DomainMessage(aggregate1Id, 0, new DomainEvent(), date);
    const event2Aggregate1 = new DomainMessage(aggregate1Id, 1, new DomainEvent(), date);
    const aggregate1Stream = SimpleDomainEventStream.of([event1Aggregate1, event2Aggregate1]);
    await repository.append(aggregate1Id, aggregate1Stream);

    const aggregate2Id = new ScalarIdentity('aggregate2 ID');
    const event1Aggregate2 = new DomainMessage(aggregate2Id, 0, new DomainEvent(), date);
    const event2Aggregate2 = new DomainMessage(aggregate2Id, 1, new DomainEvent(), date);
    const aggregate2Stream = SimpleDomainEventStream.of([event1Aggregate2, event2Aggregate2]);
    await repository.append(aggregate2Id, aggregate2Stream);

    expect(repository).toMatchSnapshot('InMemoryEventStore');
    expect(await repository.load(aggregate1Id).pipe(toArray()).toPromise()).toEqual(
      [event1Aggregate1, event2Aggregate1]);
    expect(await repository.load(aggregate2Id).pipe(toArray()).toPromise()).toEqual(
      [event1Aggregate2, event2Aggregate2]);
  });

  it('Throws an exception when playhead is not correct', async () => {
    const repository = new InMemoryEventStore();
    const id = new ScalarIdentity('38459347598437');
    const event1 = new DomainMessage(id, 0, new DomainEvent(), date);
    const event2 = new DomainMessage(id, 2, new DomainEvent(), date);
    const event3 = new DomainMessage(id, 1, new DomainEvent(), date);
    const eventStream = SimpleDomainEventStream.of([event1, event2, event3]);
    const playheadError = PlayheadError.create(1, 2);
    await expect(repository.append(id, eventStream)).rejects.toThrowError(playheadError as any);
  });

  it('Can load all', async () => {
    const store = new InMemoryEventStore();
    const id1 = new ScalarIdentity('1');
    const event1 = new DomainMessage(id1, 0, new DomainEvent(), date);
    const eventStream1 = SimpleDomainEventStream.of([event1]);
    await store.append(id1, eventStream1);

    const id2 = new ScalarIdentity('2');
    const event2 = new DomainMessage(id2, 0, new DomainEvent(), date);
    const eventStream2 = SimpleDomainEventStream.of([event2]);
    await store.append(id2, eventStream2);

    const all = store.loadAll();
    expect(await all.pipe(toArray()).toPromise()).toEqual([event1, event2]);
  });

  it('Can load from certain playhead', async () => {
    const store = new InMemoryEventStore();
    const id = new ScalarIdentity('38459347598437');
    const event1 = new DomainMessage(id, 0, new DomainEvent(), date);
    const event2 = new DomainMessage(id, 1, new DomainEvent(), date);
    const event3 = new DomainMessage(id, 2, new DomainEvent(), date);
    const event4 = new DomainMessage(id, 3, new DomainEvent(), date);
    const eventStream = SimpleDomainEventStream.of([event1, event2, event3, event4]);

    await store.append(id, eventStream);
    const fromPlayhead = store.loadFromPlayhead(id, 2);
    expect(await fromPlayhead.pipe(toArray()).toPromise()).toEqual([event3, event4]);
  });

  it('Knows it has an event stream', async () => {
    const store = new InMemoryEventStore();
    const id = new ScalarIdentity('38459347598437');
    const event1 = new DomainMessage(id, 0, new DomainEvent(), date);
    const eventStream = SimpleDomainEventStream.of([event1]);
    await store.append(id, eventStream);
    expect(await store.has(id)).toBeTruthy();
    expect(await store.has(new ScalarIdentity('does not exists'))).toBeFalsy();
  });

  it('Can create from array', async () => {
    const id = new ScalarIdentity('38459347598437');
    const event1 = new DomainMessage(id, 0, new DomainEvent(), date);
    const event2 = new DomainMessage(id, 1, new DomainEvent(), date);
    const event3 = new DomainMessage(id, 2, new DomainEvent(), date);
    const event4 = new DomainMessage(id, 3, new DomainEvent(), date);

    const store = InMemoryEventStore.fromArray([event1, event2, event3, event4]);

    expect(await store.load(id).pipe(toArray()).toPromise()).toEqual([event1, event2, event3, event4]);
  });

  it('Should wait for mutex to be finished', async () => {
    const now = new Date();
    const store = new InMemoryEventStore(() => new Mutex());
    const id = 'c2950d0e-71e6-4bf3-8389-8a98981fb67c';
    const stream1 = timer(50).pipe(take(1), map(() => new DomainMessage(id, 0, { foo: 'foo' }, now, {})));
    const stream2 = timer(100).pipe(take(1), map(() => new DomainMessage(id, 1, { bar: 'bar' }, now, {})));

    const pendingStore = store.append(id, stream1);

    // Should wait for this to complete, becuase we are in pending store.
    const load = store.load(id).pipe(toArray()).toPromise();
    const has = store.has(id);
    const hasNot = store.has(v4());
    const append = store.append(id, stream2);
    await pendingStore;

    expect(await load).toEqual([new DomainMessage(id, 0, { foo: 'foo' }, now, {})]);
    expect(await has).toEqual(true);
    expect(await hasNot).toEqual(false);

    await append;

    expect(await store.load(id).pipe(toArray()).toPromise()).toEqual([
      new DomainMessage(id, 0, { foo: 'foo' }, now, {}),
      new DomainMessage(id, 1, { bar: 'bar' }, now, {}),
    ]);

  });

});
