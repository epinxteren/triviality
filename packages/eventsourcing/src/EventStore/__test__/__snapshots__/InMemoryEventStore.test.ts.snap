// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InMemoryEventStore Can have multiple different events streams: InMemoryEventStore 1`] = `
InMemoryEventStore {
  "events": ValueMutex {
    "lastSetValue": 2,
    "mutex": Mutex {
      "_semaphore": Semaphore {
        "_currentReleaser": [Function],
        "_queue": Array [],
        "_value": 1,
      },
    },
    "value": Array [
      DomainMessage {
        "aggregateId": ScalarIdentity {
          "id": "aggregate1 ID",
        },
        "metadata": Object {},
        "payload": DomainEvent {},
        "playhead": 0,
        "recordedOn": 1970-01-18T18:26:23.762Z,
      },
      DomainMessage {
        "aggregateId": ScalarIdentity {
          "id": "aggregate1 ID",
        },
        "metadata": Object {},
        "payload": DomainEvent {},
        "playhead": 1,
        "recordedOn": 1970-01-18T18:26:23.762Z,
      },
      DomainMessage {
        "aggregateId": ScalarIdentity {
          "id": "aggregate2 ID",
        },
        "metadata": Object {},
        "payload": DomainEvent {},
        "playhead": 0,
        "recordedOn": 1970-01-18T18:26:23.762Z,
      },
      DomainMessage {
        "aggregateId": ScalarIdentity {
          "id": "aggregate2 ID",
        },
        "metadata": Object {},
        "payload": DomainEvent {},
        "playhead": 1,
        "recordedOn": 1970-01-18T18:26:23.762Z,
      },
    ],
  },
  "eventsMap": Map {
    ScalarIdentity {
      "id": "aggregate1 ID",
    } => ValueMutex {
      "lastSetValue": 1,
      "mutex": Mutex {
        "_semaphore": Semaphore {
          "_currentReleaser": [Function],
          "_queue": Array [],
          "_value": 1,
        },
      },
      "value": Array [
        DomainMessage {
          "aggregateId": ScalarIdentity {
            "id": "aggregate1 ID",
          },
          "metadata": Object {},
          "payload": DomainEvent {},
          "playhead": 0,
          "recordedOn": 1970-01-18T18:26:23.762Z,
        },
        DomainMessage {
          "aggregateId": ScalarIdentity {
            "id": "aggregate1 ID",
          },
          "metadata": Object {},
          "payload": DomainEvent {},
          "playhead": 1,
          "recordedOn": 1970-01-18T18:26:23.762Z,
        },
      ],
    },
    ScalarIdentity {
      "id": "aggregate2 ID",
    } => ValueMutex {
      "lastSetValue": 1,
      "mutex": Mutex {
        "_semaphore": Semaphore {
          "_currentReleaser": [Function],
          "_queue": Array [],
          "_value": 1,
        },
      },
      "value": Array [
        DomainMessage {
          "aggregateId": ScalarIdentity {
            "id": "aggregate2 ID",
          },
          "metadata": Object {},
          "payload": DomainEvent {},
          "playhead": 0,
          "recordedOn": 1970-01-18T18:26:23.762Z,
        },
        DomainMessage {
          "aggregateId": ScalarIdentity {
            "id": "aggregate2 ID",
          },
          "metadata": Object {},
          "payload": DomainEvent {},
          "playhead": 1,
          "recordedOn": 1970-01-18T18:26:23.762Z,
        },
      ],
    },
  },
  "mutexFactory": [Function],
}
`;
